test/net-api/test-connection.js

```diff
@@ -0,0 +1,107 @@
+var tape = require('tape')
+var utp = require('../../index.js')
+
+tape('net-api connection', function (t) {
+  t.test('connect', function (t) {
+    var connected = false
+
+    utp.createServer(function () {
+      connected = true
+    }).listen(53410)
+
+    var socket = utp.connect(53410)
+
+    socket.on('connect', function () {
+      t.ok(connected, 'connected successfully')
+      t.end()
+    })
+  })
+
+  t.test('connect deferred', function (t) {
+    var connected = false
+
+    setTimeout(function () {
+      utp.createServer(function () {
+        connected = true
+      }).listen(53400)
+    }, 100)
+
+    var socket = utp.connect(53400)
+    socket.on('connect', function () {
+      t.ok(connected, 'connected successfully')
+      t.end()
+    })
+  })
+
+  t.skip('close', function (t) {
+    var server = utp.createServer(function (socket) {
+      socket.resume()
+      socket.end()
+      // TODO callback on server.close is not implemented
+      server.close(function () {
+        t.pass('closed')
+        t.end()
+      })
+    })
+
+    server.listen(53454, function () {
+      var socket = utp.connect(53454)
+
+      socket.once('connect', function () {
+        socket.end()
+      })
+    })
+  })
+
+  t.skip('on close', function (t) {
+    var closed = 0
+
+    var onclose = function () {
+      closed++
+      if (closed === 2) {
+        t.end()
+      }
+    }
+    utp.createServer(function (socket) {
+      socket.resume()
+      socket.on('end', function () {
+        socket.end()
+      })
+      socket.on('close', onclose)
+    }).listen(53455)
+
+    var socket = utp.connect(53455)
+
+    socket.resume()
+    socket.on('close', onclose)
+    socket.on('error', function (err) {
+      // TODO fix: ending from the server side is causing a ECONNRESET to be
+      // thrown on the client
+      console.log(err)
+    })
+    socket.end()
+  })
+
+  t.test('end', function (t) {
+    var ended = false
+
+    utp.createServer(function (socket) {
+      socket.resume()
+      socket.on('end', function () {
+        ended = true
+        socket.end()
+      })
+    }).listen(53454)
+
+    var socket = utp.connect(53454)
+
+    socket.resume()
+    socket.on('end', function () {
+      // TODO fix: socket is ending before the server sock actually call .end()
+      t.ok(ended)
+      t.end()
+    })
```

Comment:
Similar to the above, however here, the server socket doesn't even reach a 'end' state, the client socket emits an 'end' as soon as we call it to end it. This is probably an expected protocol thing, just want to make sure, I'll add a note to the README.